### PR TITLE
Fixes #26805 - react-router foundation #4

### PR DESCRIPTION
**Pull request by glekner**
- add server props to every route ( are there any more props a route should have ? )
- write a helper to get searchBar Props. @Laviro please take a look
- add pagination props (disablePrev, disableNext)

### CODE REVIEW

File: webpack/assets/javascripts/react_app/routes/index.js [hunk 1]
```diff
@@ -1,23 +1,22 @@
...
+        path={path}
+        key={path}
+        {...routeProps}
         render={renderProps => {
           const railsContainer = document.getElementById('rails-app-content');
           if (railsContainer) railsContainer.remove();
           currentLocation = renderProps.location;
 
-          return render(renderProps);
+          return render({ ...renderProps, ...serverProps });
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/routes/index.js]: Does it mean only routes will receive the `metadata` props?

imo it should be an `application-metadata` reducer instead so all components get the app-metadata info by using selectors.

File: webpack/assets/javascripts/react_app/components/Pagination/Pagination.js [hunk 2]
```diff
@@ -17,6 +17,8 @@ const Pagination = props => {
     onPageSet,
     onPerPageSelect,
     dropdownButtonId,
+    disableNext,
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Pagination/Pagination.js:20]: Is this apart of the scope? Can it be done in a different PR?

**Review by sharvit — CHANGES_REQUESTED** [review]

File: webpack/assets/javascripts/react_app/components/Pagination/Pagination.js (hunk 2 above)
Comment (glekner) [reply webpack/assets/javascripts/react_app/components/Pagination/Pagination.js:20]: its part of the scope yes, adding foundation for router pagination.

File: webpack/assets/javascripts/react_app/constants.js [hunk 3]
```diff
@@ -3,3 +3,22 @@ export const STATUS = {
...
+
+export const getControllerSearchProps = (
+  controller,
+  id = 'searchBar',
+  canCreate = true
+) => ({
+  controller,
+  autocomplete: {
+    id,
+    searchQuery: '',
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/constants.js:15]: `searchQuery` should be able to update from server

File: webpack/assets/javascripts/react_app/constants.js [hunk 4]
```diff
@@ -3,3 +3,22 @@ export const STATUS = {
...
+    url: `${controller}/auto_complete_search`,
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/constants.js:16]: I think `url` should be also passed from server,
unless you would like to mimic: `send("auto_complete_search_#{auto_complete_controller_name}_path")`

File: webpack/assets/javascripts/react_app/constants.js [hunk 5]
```diff
@@ -3,3 +3,22 @@ export const STATUS = {
...
+    useKeyShortcuts: true,
+  },
+  bookmarks: {
+    url: '/api/bookmarks',
+    canCreate,
+    documentationUrl: `4.1.5Searching`,
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/constants.js:22]: won't you prefer to get the documentation `url` ready from server?
`documentation_url("4.1.5Searching")`

File: webpack/assets/javascripts/react_app/constants.js [hunk 6]
```diff
@@ -3,3 +3,22 @@ export const STATUS = {
...
+  controller,
+  autocomplete: {
+    id,
+    searchQuery: '',
+    url: `${controller}/auto_complete_search`,
+    useKeyShortcuts: true,
+  },
+  bookmarks: {
+    url: '/api/bookmarks',
+    canCreate,
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/constants.js]: maybe rename to `canCreateBookmarks` ?

**Comment by glekner** [comment]
_summarizing changes_
- add server props to every route ( are there any more props a route should have ? )
- write a helper to get searchBar Props. @Laviro please take a look
- add pagination props (disablePrev, disableNext)
- added a Reducer and Selectors for ReactApp, now contains a metadata state. The Layout adds server props to the state onMount
- added `withRenderHandler` HOC - Helps render the right component based on state (isLoading, hasError, hasData)
- added common EmptyPage, LoadingPage
- added `recompose` to package.json for future use (Lifecycle HOCs)

**Comment by glekner** [comment]
@sharvit - I prefer to leave the `withRenderHandler` like this, giving it a `onMount` arg can be confusing because we wrap it with a functional component thats inside a `PageLayout` wrapper. importing an action from that inner component and passing it to the HOC seems out of place to me.

**Review by ohadlevy — DISMISSED** [review]

File: package.json [hunk 7]
```diff
@@ -108,6 +108,7 @@
     "react-password-strength": "^2.4.0",
     "react-redux": "^5.1.1",
     "react-router-dom": "^4.3.1",
+    "recompose": "^0.30.0",
```
Comment (ohadlevy) [inline package.json]: packaging PR please?

Comment (glekner) [reply package.json]: https://github.com/theforeman/foreman-packaging/pull/3825

**Review by ohadlevy — DISMISSED** [review]

**Comment by ohadlevy** [comment]
LGTM - @sharvit any additional feedback?

File: package.json (hunk 7 above)
Comment (ohadlevy) [reply package.json]: btw where is this used?

Comment (glekner) [reply package.json]: not used yet here, only in the blocked PRs

Comment (ohadlevy) [reply package.json]: hmm.. taken from https://github.com/acdlite/recompose

```
A Note from the Author (acdlite, Oct 25 2018):
Hi! I created Recompose about three years ago. About a year after that, I joined the React team. Today, we announced a proposal for Hooks. Hooks solves all the problems I attempted to address with Recompose three years ago, and more on top of that. I will be discontinuing active maintenance of this package (excluding perhaps bugfixes or patches for compatibility with future React releases), and recommending that people use Hooks instead. Your existing code with Recompose will still work, just don't expect any new features. Thank you so, so much to @wuct and @istarkov for their heroic work maintaining Recompose over the last few years.

Read more discussion about this decision here.
```

**Comment by glekner** [comment]
- removed `recompose`
- added `HOC.js` inside routes folder, added `withRenderHandler` and `callOnMount` to it
- `callOnMount` now uses `useEffect` instead of `recompose`

**Comment by glekner** [comment]
added `callOnPopState` to `HOC.js`
assuming the component uses `withRouter`, you can use `callOnPopState` to trigger a function on browser navigation (popstate)

**Comment by glekner** [comment]
@sharvit can you review?

**Review by Ron-Lavi — COMMENTED** [review]
Looks great apart from the `getControllerSearchProps` as consts.

File: webpack/assets/javascripts/react_app/constants.js (hunk 3 above)
Comment (glekner) [reply webpack/assets/javascripts/react_app/constants.js:15]: `searchQuery` no longer gets updates from the server :) thats the point
with new views implementing the router approach, there won't be full page reloads.
If theres a full page reload, the page will fetch it from the URI on mount.

File: webpack/assets/javascripts/react_app/constants.js (hunk 4 above)
Comment (glekner) [reply webpack/assets/javascripts/react_app/constants.js:16]: same as the comment above, there are no longer props from the server for routes. do we want to avoid an API call here? @ohadlevy

File: webpack/assets/javascripts/react_app/constants.js (hunk 5 above)
Comment (glekner) [reply webpack/assets/javascripts/react_app/constants.js:22]: again, no props from the server. thats why its now in redux

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/constants.js:22]: in the last comments I meant using / implementing those methods that are used in the server.
Receiving them as `metadata` is also optional, isn't it ?

Comment (glekner) [reply webpack/assets/javascripts/react_app/constants.js:22]: instead of maintaining the functions in two places we receive it [here](https://github.com/theforeman/foreman/pull/6767/files#diff-127a850b7d7b20173ad3fba88f8a25e3R532) for now.
could be good to move the implementation to the client later

**Review by ohadlevy — DISMISSED** [review]

**Review by Ron-Lavi — DISMISSED** [review]

File: webpack/assets/javascripts/react_app/ReactApp/ReactAppHelper.js [hunk 8]
```diff
@@ -0,0 +1,5 @@
+import store from '../redux';
+import { APP_FETCH_SERVER_PROPS } from '../constants';
+
+export const initializeMetadata = metadata =>
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/ReactApp/ReactAppHelper.js]: Can't it be an action? It feels like a workaround redux when it can simply be implemented with redux best practices.

File: webpack/assets/javascripts/react_app/constants.js [hunk 9]
```diff
@@ -1,5 +1,26 @@
+export const APP_FETCH_SERVER_PROPS = 'APP_FETCH_SERVER_PROPS';
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/constants.js]: IMO it should belong to `ReactAppConstants.js`

File: webpack/assets/javascripts/react_app/routes/HOC.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
+import React, { useEffect } from 'react';
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/routes/HOC.js]: This file contains really nice HOCs :+1: 

I'm wondering how the documentation part of it should be handled, especially when there are no use cases yet.
I believe adding usage examples in the comments would do the job.
Another option is to update the storybook with some examples.
Or move it to another PR that use this infrastructure.

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @glekner, I like the direction you are taking us :+1:

**Comment by glekner** [comment]
@sharvit made ReactApp connected with actions.

About the HOCs, I already know of potential use-cases for all of them :) we should work on good documentation and deep dives.

File: webpack/assets/javascripts/react_app/ReactApp/ReactApp.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+import Layout from '../components/Layout';
+import AppSwitcher from '../routes';
+
+const ReactApp = ({
+  data: { layout, metadata },
+  initializeMetadata,
+  stateMetadata,
+}) => {
+  if (isEmpty(stateMetadata)) initializeMetadata(metadata);
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/ReactApp/ReactApp.js]: Shouldn't it happen only `onMount`?

File: webpack/assets/javascripts/react_app/ReactApp/ReactAppActions.js [hunk 12]
```diff
@@ -0,0 +1,4 @@
+import { APP_FETCH_SERVER_PROPS } from './ReactAppConstants';
+
+export const initializeMetadata = metadata => dispatch =>
+  dispatch({ type: APP_FETCH_SERVER_PROPS, payload: metadata });
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/ReactApp/ReactAppActions.js]: `dispatch` is redundant.

**Comment by sharvit** [comment]
> About the HOCs, I already know of potential use-cases for all of them :) we should work on good documentation and deep dives.

+1

Can you just add simple usage example in the comments?

File: webpack/assets/javascripts/react_app/routes/HOC.js (hunk 10 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/routes/HOC.js]: Can you write tests for the HOCs

File: webpack/assets/javascripts/react_app/ReactApp/ReactApp.js (hunk 11 above)
Comment (glekner) [reply webpack/assets/javascripts/react_app/ReactApp/ReactApp.js]: this is to prevent the action if not needed?

**Comment by glekner** [comment]
- removed `onPopState` HOC for now
- added tests for the HOCs
- using `callOnMount` on `ReactApp` 🚀

**Review by sharvit — APPROVED** [review]
LGTM 👍 Thanks @glekner 

It seems that failing tests are not relevant, but I'm not 100% sure.

**Review by ohadlevy — APPROVED** [review]
